utils/testgen.py

```diff
@@ -328,92 +259,53 @@ def pytest_generate_tests(metafunc):
         for more details.
 
     """
+    pf = ProviderFilter(types=types, required_fields=required_fields)
+    return providers(metafunc, filters=[pf])
 
-    argnames = []
-    argvalues = []
-    idlist = []
-
-    for provider in cfme_data.get('management_systems', {}):
-
-        # Check provider hasn't been filtered out with --use-provider
-        if provider not in filtered:
-            continue
 
-        try:
-            prov_obj = get_crud(provider)
-        except UnknownProviderType:
-            continue
-
-        if not prov_obj:
-            logger.debug("Whilst trying to create an object for %s we failed", provider)
-            continue
-
-        if provider_types is not None:
-            if not(prov_obj.type_tclass in provider_types or prov_obj.type_name in provider_types):
-                continue
-
-        # Run through all the testgen uncollect fns
-        uncollect = False
-        uncollect_fns = [_uncollect_restricted_version, _check_required_fields,
-            _uncollect_test_flags, _uncollect_since_version]
-        for fn in uncollect_fns:
-            if fn(prov_obj.data, metafunc, required_fields):
-                uncollect = True
-                break
-        if uncollect:
-            continue
+def providers_by_category(metafunc, categories, required_fields=None):
+    """Get the values of the named field keys from ``cfme_data.get('management_systems', {})``
+    Args:
+        categories: A list of provider categories to include. If None, all providers are considered
 
-        if 'provider' in metafunc.fixturenames and 'provider' not in argnames:
-            metafunc.function = pytest.mark.uses_testgen()(metafunc.function)
-            argnames.append('provider')
+    Usage:
 
-        # uncollect when required field is not present and option['require_field'] == True
-        argvalues.append([prov_obj])
+        # In the function itself
+        def pytest_generate_tests(metafunc):
+            argnames, argvalues, idlist = testgen.providers_by_category(
+                ['cloud', 'containers'], required_fields=['provisioning']
+            )
+        metafunc.parametrize(argnames, argvalues, ids=idlist, scope='module')
 
-        # Use the provider name for idlist, helps with readable parametrized test output
-        idlist.append(provider)
+        # Using the parametrize wrapper
+        pytest_generate_tests = testgen.parametrize(testgen.providers_by_category, ['cloud'],
+            scope='module')
 
-    return argnames, argvalues, idlist
+    Note:
+        See :py:func:`providers_by_type` for more information.
+    """
+    pf = ProviderFilter(categories=categories, required_fields=required_fields)
+    return providers(metafunc, filters=[pf])
 
 
 def cloud_providers(metafunc, **options):
```

Comment:
I want to get rid of these. We shouldn't use them they are far too specific to provider categories. Thoughts?